### PR TITLE
DOPE-255: add let clause with tests

### PR DESCRIPTION
**Pull request by jansigi**

### CODE REVIEW

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    fun toDefinitionDopeQuery(manager: DopeQueryManager): DopeQuery {
+        val typeExpressionDopeQuery = expression.toDopeQuery(manager)
+        return DopeQuery(
+            queryString = "`$alias` = ${typeExpressionDopeQuery.queryString}",
+            parameters = typeExpressionDopeQuery.parameters,
+        )
+    }
+}
+
+fun <T : ValidType> String.assignTo(expression: TypeExpression<T>): LetExpression<T> = LetExpression(this, expression)
```
Comment (jansigi) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt:49]: I don't really know how I should name this function.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/ISelectClause.kt [hunk 2]
```diff
@@ -133,6 +135,12 @@ interface ISelectUnnestClause<T : ValidType> : ISelectJoinClause<T> {
     fun <U : ValidType> unnest(arrayField: Field<ArrayType<U>>) = UnnestClause(arrayField, this)
     fun <U : ValidType> unnest(aliasedArrayExpression: AliasedExpression<ArrayType<U>>) =
         AliasedUnnestClause(aliasedArrayExpression, this)
+
+    fun let(letExpression: LetExpression<out ValidType>, vararg letExpressions: LetExpression<out ValidType>) = LetClause(
+        letExpression,
+        *letExpressions,
+        parentClause = this,
+    )
```
Comment (martinagallati-ergon) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/ISelectClause.kt:143]: In the couchbase syntax, the LET clause needs to be after the JOIN clause.

**Review by martinagallati-ergon — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        val letExpressionDopeQueries = letExpressions.map { it.toDefinitionDopeQuery(manager) }
+
+        return DopeQuery(
+            queryString = formatToQueryStringWithSymbol(
+                parentDopeQuery.queryString,
+                "LET",
+                letExpressionDopeQuery.queryString,
+                *letExpressionDopeQueries.map { it.queryString }.toTypedArray(),
+            ),
+            parameters = parentDopeQuery.parameters.merge(*letExpressionDopeQueries.map { it.parameters }.toTypedArray()),
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt:28]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    fun toDefinitionDopeQuery(manager: DopeQueryManager): DopeQuery {
+        val typeExpressionDopeQuery = expression.toDopeQuery(manager)
+        return DopeQuery(
+            queryString = "`$alias` = ${typeExpressionDopeQuery.queryString}",
+            parameters = typeExpressionDopeQuery.parameters,
+        )
+    }
+}
+
+fun <T : ValidType> String.assignTo(expression: TypeExpression<T>): LetExpression<T> = LetExpression(this, expression)
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt:51]: We should definitely support it for non-dopetypes as well (`assignTo(5)`)

File: core/src/test/kotlin/ch/ergon/dope/buildTest/QueryBuilderTest.kt [hunk 5]
```diff
@@ -896,4 +901,30 @@ class QueryBuilderTest : ManagerDependentTest {
 
         assertEquals(expected, actual)
     }
+
+    @Test
+    fun `should support selecting let clause`() {
```
Comment (pgruntz) [inline core/src/test/kotlin/ch/ergon/dope/buildTest/QueryBuilderTest.kt:906]: I think we should define the let-clause in the following way:

```
val var1 = DopeVariable("var1", 5)
val var2 = DopeVariable("var2", "lel")
val variables = listOf(var1, var2)

create.select(...)
.from(...)
.withVariables(variables)
.where(var1.isEqualTo(7) //Typecheck
.and(var2.isEqualTo(7)) //Compiler error
```

Comment (pgruntz) [reply core/src/test/kotlin/ch/ergon/dope/buildTest/QueryBuilderTest.kt:906]: Better would be if the user doesn't need to pass the variables in the _.withVariables_ method but Dope will collect all variables used in the sub-clauses and add it in the let-part

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/test/kotlin/ch/ergon/dope/buildTest/QueryBuilderTest.kt (hunk 5 above)
Comment (jansigi) [reply core/src/test/kotlin/ch/ergon/dope/buildTest/QueryBuilderTest.kt:906]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt (hunk 4 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt:51]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt (hunk 1 above)
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt]: I will keep the name assignTo

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import ch.ergon.dope.validtype.ValidType
+
+class LetClause<T : ValidType>(
+    private val dopeVariable: DopeVariable<out ValidType>,
+    private vararg val dopeVariables: DopeVariable<out ValidType>,
+    private val parentClause: ISelectLetClause<T>,
+) : ISelectLetClause<T> {
+    override fun toDopeQuery(manager: DopeQueryManager): DopeQuery {
+        val parentDopeQuery = parentClause.toDopeQuery(manager)
+        val letExpressionDopeQuery = dopeVariable.toDefinitionDopeQuery(manager)
```
[the remark this answers is not in the document]
Comment (jansigi) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/LetClause.kt:21]: done

**Review by martinagallati-ergon — APPROVED** [review]
Nice, I like this approach! Because it's a new clause, I would consider adding an integration test; the build test you wrote seems like a good candidate for that. What do you think?

**Review by pgruntz — APPROVED** [review]

**Pull request merged**